python/pyspark/sql/dataframe.py

```diff
@@ -1272,16 +1272,18 @@ def replace(self, to_replace, value, subset=None):
         """Returns a new :class:`DataFrame` replacing a value with another value.
         :func:`DataFrame.replace` and :func:`DataFrameNaFunctions.replace` are
         aliases of each other.
+        Values `to_replace` and `value` should be homogeneous. Mixed string and numeric
```

Comment:
I've seen some variation between bundled Py4j (`py4j-0.10.4-src.zip`) and pip installed version (should be the same `py4j==0.10.4`) which suggests there is something more going on here but I would really avoid restricting Python users even more. 

If necessary we can always proactively cast all `ints` to `floats`. It will be an ugly hack but won't affect user code or correctness.